### PR TITLE
ci: added publishing intel trust authority AS docker

### PR DESCRIPTION
**Pull request by pawelpros**
- Refactored directory structure for building KBS docker images
- Added publishing KBS intel trust authority AS docker image on ghcr.io

### CODE REVIEW

**Comment by fitzthum** [comment]
DCO check complains because the author does not match the sign-off. Basically it should match `git config --local user.name `. You can update that config and then do `git commit --amend --reset-author` I think.

**Comment by fitzthum** [comment]
You should probably update the [release script](https://github.com/confidential-containers/trustee/blob/main/hack/release-helper.sh) to make sure your staged image gets copied over as a release image.

**Comment by pawelpros** [comment]
Thanks for the tip - should be ok right now - I have also aligned with recent changes

File: hack/release-helper.sh [hunk 1]
```diff
@@ -9,6 +9,7 @@ declare -g release_tag
 declare -A staged_to_release=(
     ["staged-images/kbs"]="key-broker-service"
     ["staged-images/kbs-grpc-as"]="key-broker-service"
+    ["staged-images/kbs-ita-as"]="key-broker-service"
```
Comment (fitzthum) [inline hack/release-helper.sh:12]: This part remaps the image names, but I think you'll also need to create an entry below where the tag prefix is generated. Otherwise, we're going to have a collision with the image names. We should maybe add some comments to the script to make it more clear.

**Review by fitzthum — COMMENTED** [review]
Looks fine. I don't see anything bumping OpenSSL

Also note that we use the same naming convention for Dockerfiles in the AS dir. Do we care about this being consistent?

File: hack/release-helper.sh (hunk 1 above)
Comment (pawelpros) [reply hack/release-helper.sh:14]: Fixed along with release-guide + comment (which is helpful in this case)

File: hack/release-helper.sh [hunk 2]
```diff
@@ -6,15 +6,20 @@ declare -g gh_username
...
     ["staged-images/rvps"]="reference-value-provider-service"
     ["staged-images/coco-as-grpc"]="attestation-service"
     ["staged-images/coco-as-restful"]="attestation-service"
 )
 declare -A staged_to_release_tag_prefix=(
     ["staged-images/kbs"]="built-in-as-"
+    ["staged-images/kbs-grpc-as"]="grpc-as-"
```
[the remark this answers is not in the document]
Comment (pawelpros) [reply hack/release-helper.sh:21]: No problem - I'll revert this line from commit and leave only second one

File: .github/workflows/as-build-and-push.yaml [hunk 3]
```diff
@@ -102,11 +102,11 @@ jobs:
...
-          --amend ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:latest-x86_64
-        docker manifest push ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:latest
+        docker manifest create "ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:${commit_sha}" \
+          --amend "ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:${commit_sha}-s390x" \
+          --amend "ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:${commit_sha}-x86_64"
+        docker manifest push "ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:${commit_sha}"
+        docker manifest create "ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:latest" \
+          --amend "ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:latest-s390x" \
+          --amend "ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:latest-x86_64"
+        docker manifest push "ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:latest"
```
Comment (fidencio) [inline .github/workflows/as-build-and-push.yaml:112]: I'm calling the quoting "Change 2", and this can go on its own commit (as part of the very same PR).

File: .github/workflows/kbs-build-and-push.yaml [hunk 4]
```diff
@@ -57,10 +62,10 @@ jobs:
...
         https_crypto=${{ matrix.https_crypto }}
         [ "${arch}" = "s390x" ] && https_crypto=openssl
-        DOCKER_BUILDKIT=1 docker build -f ${{ matrix.docker_file }} --push \
-          -t ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:${commit_sha}-${arch} \
-          -t ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:latest-${arch} \
-          --build-arg ARCH=${arch} --build-arg HTTPS_CRYPTO=${https_crypto} .
+        DOCKER_BUILDKIT=1 docker build -f "${{ matrix.docker_file }}" --push \
+          -t "ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:${commit_sha}-${arch}" \
+          -t "ghcr.io/confidential-containers/staged-images/${{ matrix.tag }}:latest-${arch}" \
+          --build-arg ARCH="${arch}" --build-arg HTTPS_CRYPTO="${https_crypto}" .
```
Comment (fidencio) [inline .github/workflows/kbs-build-and-push.yaml:68]: Change 2.

File: .github/workflows/kbs-build-and-push.yaml [hunk 5]
```diff
@@ -93,11 +99,11 @@ jobs:
...
-          --amend ghcr.io/confidential-containers/staged-images/${{ matrix.image }}:latest-s390x
-        docker manifest push ghcr.io/confidential-containers/staged-images/${{ matrix.image }}:latest
+        docker manifest create "ghcr.io/confidential-containers/staged-images/${{ matrix.image }}:${commit_sha}" \
+          --amend "ghcr.io/confidential-containers/staged-images/${{ matrix.image }}:${commit_sha}-x86_64" \
+          --amend "ghcr.io/confidential-containers/staged-images/${{ matrix.image }}:${commit_sha}-s390x"
+        docker manifest push "ghcr.io/confidential-containers/staged-images/${{ matrix.image }}:${commit_sha}"
+        docker manifest create "ghcr.io/confidential-containers/staged-images/${{ matrix.image }}:latest" \
+          --amend "ghcr.io/confidential-containers/staged-images/${{ matrix.image }}:latest-x86_64" \
+          --amend "ghcr.io/confidential-containers/staged-images/${{ matrix.image }}:latest-s390x"
+        docker manifest push "ghcr.io/confidential-containers/staged-images/${{ matrix.image }}:latest"
```
Comment (fidencio) [inline .github/workflows/kbs-build-and-push.yaml:109]: Change 2.

**Review by fidencio — COMMENTED** [review]
Change 2 is good, but I'd recommend doing it as part of its own commit(s).

File: .github/workflows/as-dockerbuild.yml [hunk 6]
```diff
@@ -28,12 +28,12 @@ jobs:
...
     - name: Build RESTful AS Container Image
       run: |
-        DOCKER_BUILDKIT=1 docker build -t attestation-service:latest . -f attestation-service/Dockerfile.as-restful
+        DOCKER_BUILDKIT=1 docker build -t attestation-service:latest . -f attestation-service/docker/as-restful/Dockerfile
 
     - name: Build RVPS Container Image
       run: |
-        Docker_BUILDKIT=1 docker build -t rvps:latest . -f attestation-service/rvps/Dockerfile
\ No newline at end of file
+        Docker_BUILDKIT=1 docker build -t rvps:latest . -f attestation-service/docker/rvps/Dockerfile
```
Comment (fidencio) [inline .github/workflows/as-dockerbuild.yml:39]: Change 1.

File: .github/workflows/kbs-build-and-push.yaml [hunk 7]
```diff
@@ -18,15 +18,20 @@ jobs:
         tag:
           - kbs
           - kbs-grpc-as
+          - kbs-ita-as
```
Comment (fidencio) [inline .github/workflows/kbs-build-and-push.yaml:21]: Change 3, ITA!

File: .github/workflows/kbs-build-and-push.yaml [hunk 8]
```diff
@@ -78,6 +83,7 @@ jobs:
         image:
           - kbs
           - kbs-grpc-as
+          - kbs-ita-as
```
Comment (fidencio) [inline .github/workflows/kbs-build-and-push.yaml:86]: Change 3.

File: .github/workflows/kbs-docker-build.yml [hunk 9]
```diff
@@ -16,5 +16,6 @@ jobs:
       run: |
         DOCKER_BUILDKIT=1 docker build -t kbs:coco-as . -f kbs/docker/Dockerfile; \
         DOCKER_BUILDKIT=1 docker build -t kbs:coco-as-openssl --build-arg KBS_FEATURES=coco-as-builtin,openssl,resource,opa . -f kbs/docker/Dockerfile; \
-        DOCKER_BUILDKIT=1 docker build -t kbs:coco-as-grpc . -f kbs/docker/Dockerfile.coco-as-grpc; \
-        DOCKER_BUILDKIT=1 docker build -t kbs:coco-as-rhel-ubi . -f kbs/docker/Dockerfile.rhel-ubi
+        DOCKER_BUILDKIT=1 docker build -t kbs:coco-as-grpc . -f kbs/docker/coco-as-grpc/Dockerfile; \
+        DOCKER_BUILDKIT=1 docker build -t kbs:coco-as-rhel-ubi . -f kbs/docker/rhel-ubi/Dockerfile; \
+        DOCKER_BUILDKIT=1 docker build -t kbs:coco-as-intel-trust-authority . -f kbs/docker/intel-trust-authority/Dockerfile
```
Comment (fidencio) [inline .github/workflows/kbs-docker-build.yml:21]: This part here is a mix of Change 1, 2, and 3.

File: hack/release-helper.sh [hunk 10]
```diff
@@ -6,15 +6,19 @@ declare -g gh_username
 declare -g gh_token
 declare -g release_candidate_sha
 declare -g release_tag
+
+# Output naming convention along with release guide can be found in release-guide.md
 declare -A staged_to_release=(
     ["staged-images/kbs"]="key-broker-service"
     ["staged-images/kbs-grpc-as"]="key-broker-service"
+    ["staged-images/kbs-ita-as"]="key-broker-service"
```
Comment (fidencio) [inline hack/release-helper.sh:14]: Change 3

File: hack/release-helper.sh [hunk 11]
```diff
@@ -6,15 +6,19 @@ declare -g gh_username
...
     ["staged-images/rvps"]="reference-value-provider-service"
     ["staged-images/coco-as-grpc"]="attestation-service"
     ["staged-images/coco-as-restful"]="attestation-service"
 )
 declare -A staged_to_release_tag_prefix=(
     ["staged-images/kbs"]="built-in-as-"
+    ["staged-images/kbs-ita-as"]="ita-as-"
```
Comment (fidencio) [inline hack/release-helper.sh:21]: Change 3.

File: kbs/docker/intel-trust-authority/Dockerfile [hunk 12]
```diff
@@ -1,12 +1,14 @@
...
+ARG HTTPS_CRYPTO=rustls
 
 WORKDIR /usr/src/kbs
 COPY . .
 
 RUN apt-get update && apt install -y git
 
 # Build and Install KBS
-RUN cargo install --path kbs/src/kbs --no-default-features --features intel-trust-authority-as,rustls,resource,opa
+RUN cargo install --path kbs/src/kbs --no-default-features --features intel-trust-authority-as,${HTTPS_CRYPTO},resource,opa
```
Comment (fidencio) [inline kbs/docker/intel-trust-authority/Dockerfile:11]: Change 3.

File: release-guide.md [hunk 13]
```diff
@@ -16,7 +16,8 @@ ghcr, and assuming you are releasing v0.8.2, these would be the ghcr package
 mappings:
 ```
 staged-images/kbs:latest -> key-broker-service:built-in-as-v0.8.2
-staged-images/kbs-grpc-as:latest -> key-broker-service:v0.8.2
+staged-images/kbs-grpc-as:latest -> key-broker-service:grpc-as-v0.8.2
+staged-images/kbs-ita-as:latest -> key-broker-service:ita-as-v0.8.2
```
Comment (fidencio) [inline release-guide.md:20]: Change 3.

File: release-guide.md [hunk 14]
```diff
@@ -16,7 +16,8 @@ ghcr, and assuming you are releasing v0.8.2, these would be the ghcr package
 mappings:
 ```
 staged-images/kbs:latest -> key-broker-service:built-in-as-v0.8.2
-staged-images/kbs-grpc-as:latest -> key-broker-service:v0.8.2
+staged-images/kbs-grpc-as:latest -> key-broker-service:grpc-as-v0.8.2
```
Comment (fidencio) [inline release-guide.md:19]: This is a change on its own.

**Review by fidencio — COMMENTED** [review]
Now, about the change 3, which also deserves to be part of its own commit (still on this very same PR).

We're adding the possibility to build the ITA image, but ... correct me if I'm mistaken ... I think we're doing multi-arch build for all the images.  With this in mind, an obvious question is ... is ITA supposed to support multi-arch? Or is it x86_64 specific?

Depending on the answer, we may have to limit this to only be built with for x86_64.

File: .github/workflows/kbs-build-and-push.yaml [hunk 15]
```diff
@@ -93,11 +93,35 @@ jobs:
...
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Login to GHCR Container Registry
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
```
Comment (Xynnn007) [inline .github/workflows/kbs-build-and-push.yaml:120]: I think these parts are duplicated. Could we only add the `docker build ... --push` partsright under the existing parts?

Comment (pawelpros) [reply .github/workflows/kbs-build-and-push.yaml:120]: I agree but ITA for now shouldn't be automatically built for arch different than x86_64 and with current approach is not clear for me how to approach this without if statements and running workflow twice (for both arch types)

Comment (pawelpros) [reply .github/workflows/kbs-build-and-push.yaml]: Excluded one config and added to exiting approach

**Review by fidencio — APPROVED** [review]
lgtm, thanks @pawelpros!

**Review by fitzthum — APPROVED** [review]
LGTM
